src/modsim2/similarity/embeddings.py

```diff
@@ -19,4 +25,131 @@ def array_to_matrix(array: np.ndarray) -> np.ndarray:
     return matrix
 
 
-EMBEDDING_FN_DICT = {"matrix": array_to_matrix}
+def umap(array: np.ndarray, random_seed: int, n_components: int) -> np.ndarray:
+    """
+    UMAP embedding of array
+
+    Args:
+        array: A numpy array of data to be embedded with observations on the first index
+        random_seed: An integer to set the random seed of the UMAP algorithm
+        n_components: The number of components to return in the embedding
+
+    Returns:
+        embed_array: A numpy array of the embedded data
+    """
+    reducer = UMAP(random_state=random_seed, n_components=n_components)
+    reducer.fit(array)
+    embed_array = reducer.transform(array)
+    return embed_array
+
+
+def inception(array: np.ndarray) -> np.ndarray:
+    """
+    Embedding using the Inception V3 neural network with pre-trained weights
+
+    Args:
+        array: A numpy array of data to be embedded with observations on the first index
+
+    Returns:
+        embed_array: A numpy array of the embedded data
+    """
+
+    # Load the inception_v3 pretrained model
+    model = models.inception_v3(weights=models.Inception_V3_Weights.DEFAULT)
+    # Replace the last fully connected layer with an identity layer
+    model.fc = torch.nn.Identity()
+    # Set model to evaluation mode
+    model.eval()
+
+    # Create process to transform data to form required by inception v3
+    preprocess = transforms.Compose(
+        [
+            transforms.Resize(299),
+            transforms.CenterCrop(299),
+            transforms.Normalize(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225]),
+        ]
+    )
+
+    # Process the data to create valid input data for the model
+    input_data = preprocess(torch.Tensor(array))
+
+    # move the input and model to GPU for speed, if available
+    if torch.cuda.is_available():
+        input_data = input_data.to("cuda")
+        model.to("cuda")
+
+    # run the data input data through the model in batches
+    output_data = []
```

Comment:
if we know how many results we'll expect then we can avoid doing appends to lists (which are expensive). We can allocate an array of suitable length and add entries to it:
```
output_data = np.zeros(n_samples, n_feats)
...
output_data[batch_start: batch_end] = batch_results
```

Reply:
Have switched to slicing an np array